File: Managed/Leftice.Runtime/CoreUObject/BooleanProperty.cs
// Copyright (c) NextTurn.
// See the LICENSE.TXT file in the project root for more information.

using System;
using Leftice;
using Leftice.Processors;

namespace Unreal
{
    public sealed class BooleanProperty : Property, IProperty<bool>
    {
        internal BooleanProperty(IntPtr pointer) : base(pointer) { }

        public bool GetValue(Object @object, int index = 0) =>
            NativeMethods.GetValue(this.pointer, GetPointerOrThrow(@object),
                (uint)index < (uint)this.ArrayLength ? index : throw new ArgumentOutOfRangeException(nameof(index)));

        public void SetValue(Object @object, bool value, int index = 0) =>
            NativeMethods.SetValue(this.pointer, GetPointerOrThrow(@object), value,
                (uint)index < (uint)this.ArrayLength ? index : throw new ArgumentOutOfRangeException(nameof(index)));

        private static new class NativeMethods
        {
            [Calli]
            public static extern bool GetValue(IntPtr property, IntPtr @object, int index);

            [Calli]
            public static extern void SetValue(IntPtr property, IntPtr @object, bool value, int index);
        }
    }
}

File: Source/LefticeRuntime/Private/CoreUObject/BooleanProperty.cpp
// Copyright (c) NextTurn.
// See LICENSE.txt in the project root for more information.

#include "LefticeRuntime/Private/Exports.hpp"

#include "LefticeRuntime/Public/Compiler.hpp"

#include "Runtime/Core/Public/CoreMinimal.h"

#include "Runtime/CoreUObject/Public/UObject/UnrealType.h"

ENABLE_WARNINGS;

#define TYPE_EXPORT_MANAGED_TYPE_NAME BooleanProperty

TYPE_EXPORT_START

bool EXPORT_CALL_CONV GetValue(const UBoolProperty& Property, const UObject* Object, int32_t Index)
{
	return Property.GetPropertyValue_InContainer(Object, Index);
}

void EXPORT_CALL_CONV SetValue(const UBoolProperty& Property, UObject* Object, bool Value, int32_t Index)
{
	Property.SetPropertyValue_InContainer(Object, Value, Index);
}

const FMemberSymbol Members[] =
{
	EXPORT_METHOD(GetValue),
	EXPORT_METHOD(SetValue),
};

TYPE_EXPORT_END

#undef TYPE_EXPORT_MANAGED_TYPE_NAME

RESTORE_WARNINGS;
